[HUDI-115] Adding DefaultHoodieRecordPayload to honor ordering with combineAndGetUpdateValue

nsivabalan opened the pull request:
## What is the purpose of the pull request

Existing OverwriteWithLatestAvroPayload always chooses incoming record when combineAndGetUpdateValue is called, but there are chances a record is late arriving and needs to be ignored. Hence adding DefaultHoodieRecordPayload to honor ordering with combineAndGetUpdateValue. 
- To achieve this, have introduced new apis in HoodieRecordPayload interface by adding a new arg for Properties for all apis and deprecating existing apis. Default impl is added so that all existing implementations are compatible w/o any changes. 
- Introduced a new config called "hoodie.payload.ordering.field" in HoodieWriteClient which is set to the preCombine field value. 
So, when a user configures payload class as the new class i.e. DefaultHoodieRecordPayload, then combineAndGetUpdateValue will combine based on preCombine field with existing record in storage. 
- Added HoodiePayloadConfig to hold all payload configs going forward.

First attempt: https://github.com/apache/hudi/pull/1704

## Brief change log
- Added new apis to HoodieRecordPayload interface by adding Properties as arg to all apis and deprecating existing apis. 
- Adding DefaultHoodieRecordPayload to honor ordering with combineAndGetUpdateValue. 
- Introduced a new config called "hoodie.payload.ordering.field" in HoodieWriteClient which is set to the preCombine field value. 
- Added HoodiePayloadConfig to hold all payload configs going forward.

## Verify this pull request

This change added tests and can be verified as follows:

  - *Added tests tp TestDataSourceDefaults to verify the change.*
  - *Added TestDefaultHoodieRecordPayload to verify the new payload class*

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

nsivabalan commented:
@vinothchandar : can you take a look at this patch. I am yet to find a good name for the class. Ignoring that for now, basically I have introduced a new config called "hoodie.datasource.honor.ordering.while.merging". Based on the value set, either existing OverwriteWithLatestAvroPayload will be used or the new class will be used while instantiating HoodiePayloads. By this way, callers of combineAndGetUpdateValue don't need to pass around the properties.

garyli1019 commented on hudi-common/src/main/java/org/apache/hudi/common/model/OverwriteWithLatestAvroPayloadV1.java:
```diff
@@ -0,0 +1,120 @@
+public class OverwriteWithLatestAvroPayloadV1 extends BaseAvroPayload
+    implements HoodieRecordPayload<OverwriteWithLatestAvroPayloadV1> {
+
+  private Map<String, String> props;
```
should we use a string to save storage? Looks like this will create a map for every record.

nsivabalan replied:
yeah. lets wait to hear from @vinothchandar . Previous impl was sending in this map as arg to combineAndGetUpdateValue rather than storing it as instance var. may be thats better in terms of storage cost. I thought we could avoid making changes to all callers of combineAndGetUpdateValue.

garyli1019 commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -121,6 +121,10 @@
   private static final String DEFAULT_MERGE_DATA_VALIDATION_CHECK_ENABLED = "false";
 
+  // payload ordering field
+  private static final String PAYLOAD_ORDERING_FIELD_PROP = "hoodie.payload.ordering.field";
```
Is there any reason we need a different config from the pre-combine field? IIUC we wanna make the merging logic consistent, so `preCombine` and `combineAndGetUpdateValue` should produce the same result?

garyli1019 commented on hudi-common/src/main/java/org/apache/hudi/common/model/OverwriteWithLatestAvroPayloadV1.java:
```diff
@@ -0,0 +1,120 @@
+ * 1. preCombine - Picks the latest delta record for a key, based on an ordering field 2.
+ * combineAndGetUpdateValue/getInsertValue - Simply overwrites storage with latest delta record
+ */
+public class OverwriteWithLatestAvroPayloadV1 extends BaseAvroPayload
```
naming is hard... how about `CombineWithLargestOrderingValPayload`?

nsivabalan replied:
yeah. Ideally existing one should have been OverwriteWithIncomingAvroPayload and this new one should be OverwriteWithLatestAvroPayload.

garyli1019 commented on hudi-spark/src/main/scala/org/apache/hudi/DataSourceOptions.scala:
```diff
@@ -205,6 +205,12 @@ object DataSourceWriteOptions {
+   * record in storage.
+   */
+  val HONOR_ORDERING_WHILE_MERGING_OPT_KEY = "hoodie.datasource.honor.ordering.while.merging"
+  val DEFAULT_HONOR_ORDERING_WHILE_MERGING_OPT_VAL = "false"
```
keep this `true` is more rational for me. From a new user perspective, I don't really understand the difference between `preCombine` and `combineWithLatestValue`. It took me a while to figure this out and originally I thought this is a bug. But we need to think about the impact of the existing user.

vinothchandar replied:
I really don't like having these payload specific configs, top level in datasource options.

nsivabalan commented:
@vinothchandar : can you take a high level pass and let me know if this looks good. or do we need to have the map passed in as an arg to combineAndGetUpdateValue.

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -154,9 +154,16 @@ private[hudi] object HoodieSparkSqlWriter {
-              DataSourceUtils.createHoodieRecord(gr,
-                orderingVal, keyGenerator.getKey(gr),
-                parameters(PAYLOAD_CLASS_OPT_KEY))
+              val honorOrderingWhileMerging = parameters(HONOR_ORDERING_WHILE_MERGING_OPT_KEY).toBoolean
```
this is more and more overhead for the user to learn one more parameter when using one specific payload

nsivabalan commented:
@vinothchandar @garyli1019 : I did some revamp based on feedback. Patch is ready to be reviewed again. 
Open to naming the new class differently. Also, will fix java docs for new class based on it. As of now, it says default payload class used in deltastreamer.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/DefaultHoodieRecordPayload.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.Properties;
+
+/**
+ * Default payload used for delta streamer.
```
this is not just for delta streamer.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/DefaultHoodieRecordPayload.java:
```diff
@@ -0,0 +1,123 @@
+ * 1. preCombine - Picks the latest delta record for a key, based on an ordering field
+ * 2. combineAndGetUpdateValue/getInsertValue - Chooses the latest record based on ordering field value.
+ */
+public class DefaultHoodieRecordPayload extends BaseAvroPayload
```
is this class pretty much a modified version of `OverwriteWithLatestAvroPayload`?  if so, can we reuse some code by having that extend from this, and override methods as needed?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/DefaultHoodieRecordPayload.java:
```diff
@@ -0,0 +1,123 @@
+     * NOTE: Deletes sent via EmptyHoodieRecordPayload and/or Delete operation type do not hit this code path
+     * and need to be dealt with separately.
+     */
+    Object persistedOrderingVal = getNestedFieldVal((GenericRecord) currentValue, properties.getProperty(HoodiePayloadProps.PAYLOAD_ORDERING_FIELD_PROP), true);
```
just do the `Comparable` cast here itself?

nsivabalan replied:
I thought you had already reviewed previous PR and hence just copied as is. Will fix it.

nsivabalan replied:
actually we can't do that. it could be null.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/DefaultHoodieRecordPayload.java:
```diff
@@ -0,0 +1,123 @@
+
+    // Null check is needed here to support schema evolution. The record in storage may be from old schema where
+    // the new ordering column might not be present and hence returns null.
+    if (persistedOrderingVal != null && ((Comparable) persistedOrderingVal).compareTo(incomingOrderingVal) > 0) {
```
nts: the check means persistedOrderingVal > incomingOrderingVal i.e we should retain the value on disk

nsivabalan replied:
sorry, you wanted to confirm this behavior? can you please clarify.

vinothchandar replied:
yes . nts = note to self. sorry :/

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -29,47 +29,93 @@
-   * When more than one HoodieRecord have the same HoodieKey, this function combines them before attempting to
-   * insert/upsert (if combining turned on in HoodieClientConfig).
+   * When more than one HoodieRecord have the same HoodieKey, this function combines them before attempting to insert/upsert (if combining turned on
+   * in HoodieClientConfig).
```
please remove "if combinng turned on ..." comment. we should ideally refrain from referring to higher level constructs from lower layers.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -29,47 +29,93 @@
   /**
-   * This methods lets you write custom merging/combining logic to produce new values as a function of current value on
-   * storage and whats contained in this object.
+   * When more than one HoodieRecord have the same HoodieKey, this function combines them before attempting to insert/upsert (if combining turned on
```
same here

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -29,47 +29,93 @@
    * <p>
-   * eg: 1) You are updating counters, you may want to add counts to currentValue and write back updated counts 2) You
-   * may be reading DB redo logs, and merge them with current image for a database row on storage
+   * eg: 1) You are updating counters, you may want to add counts to currentValue and write back updated counts 2) You may be reading DB redo logs,
```
place 2 on a newline?

vinothchandar replied:
close the `</p>`

vinothchandar replied:
I think we should just refer to the other method here in the javadoc, instead of repeating the entire description here again.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -29,47 +29,93 @@
-   * using the metadata in the context of a write success or failure.
+   * Generates an avro record out of the given HoodieRecordPayload, to be written out to storage. Called when writing a new value for the given
+   * HoodieKey, wherein there is no existing record in storage to be combined against. (i.e insert) Return EMPTY to skip writing this record.
+   * This method takes in a property map as an arg so that implementation can decide their business logic based on some properties set.
```
this line can be bit pithy?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieRecordPayload.java:
```diff
@@ -29,47 +29,93 @@
+  }
+
+  /**
+   * This method can be used to extract some metadata from HoodieRecordPayload. The metadata is passed to {@code WriteStatus.markSuccess()} and
```
thanks for fixing this

vinothchandar reviewed (commented):
Left some comments. high level, lgtm. 

IIUC, we are just adding the payload without actually making it the default, right? I wonder if we should have a follow up JIRA in 0.8.0 to do this, with the necessary considerations

nsivabalan commented:
yes, you are right. we are **NOT** making this new payload default for now. Have created https://issues.apache.org/jira/browse/HUDI-1464 to track this. 
btw, have addressed all feedback.

vinothchandar commented:
@nsivabalan could you check the build?

vinothchandar reviewed (approved):
Looks good. 

cc @bhasudha
